feat: support deepseek-coder LLM

jcraftsman opened the pull request:
#### Reference Issues/PRs
Fixes #637 

#### What does this implement/fix? 
In this PR, I implemented the changes suggested in issue #637, which involves adding support for DeepSeek Coder as a model option in SWE-agent. The following modifications were made:

1. Model Integration: Added DeepSeek Coder to the list of selectable models in SWE-agent.
2. I included two minor changes to enhance the development workflow:

  - ~~Added a justfile: This file includes helpful commands for starting the server, frontend, and switching between DeepSeek and OpenAI APIs.~~ _(EDIT: I reverted back this change. justfile was not introduced in this PR)_ 
  - Updated .gitignore: Simplified the switch between preconfigured keys*.cfg files by modifying the pattern to ignore all keys*.cfg files instead of just keys.cfg.

I tested these changes and it seems that DeepSeek Coder works seamlessly with existing SWE-agent functionalities.

I hope you find this small change useful. I'm available if you have any questions, and I really enjoy the work you’re doing with this project.

Thanks!

ofirpress commented:
Hi thanks for your contribution:
1. Why do we need to make a new file for the deepseek keys? can't we just use the existing keys.cfg?
2. also not sure what that justfile thing is for, maybe Kilian will get it

jcraftsman commented:
Hi @ofirpress ,

Thanks for your feedback!

1. Initially, my changes made the environment variables `OPENAI_API_KEY` and `OPENAI_API_BASE_URL` applicable to both OpenAI and DeepSeek models, which required manual updates in keys.cfg when switching between them. I've now introduced `DEEPSEEK_API_KEY` and `DEEPSEEK_API_BASE_URL` for DeepSeek in the [latest commit (8941058)](https://github.com/princeton-nlp/SWE-agent/pull/638/commits/8941058a55c2a3620343931d37aa5ee560f8ec1a), eliminating the need for manual changes.
2. I added a [justfile](https://just.systems/man/en/) to streamline my workflow (e.g., `just start-server`, `just start-frontend`) since `./start_web_ui.sh` didn’t work for me. This change is [isolated in commit ed9ec299](https://github.com/princeton-nlp/SWE-agent/pull/638/commits/ed9ec299ea7b2e75dc6df92ece9e3c8fa6498e4d) and can be easily removed if too specific.
The updated .gitignore simplifies managing multiple preconfigured keys (e.g., keys.projectname.cfg).

I hope this clarifies my intent. Open to any further suggestions!

jcraftsman commented:
I just dropped the justfile to avoid any confusion ^^

klieret commented:
Thanks for opening the PR. I'll take a look at this in the next few days

klieret commented:
Could you tick the "Allow edits from maintainers" checkbox: 
![image](https://github.com/princeton-nlp/SWE-agent/assets/13602468/7d7c925c-d9b0-4e6a-a83b-e0c4b4116eca)

I want to slightly refactor this

jcraftsman commented:
Hi @klieret,

I don't have access to the "Allow edits from maintainers" checkbox for this PR, likely because the changes are from a forked repo. However, I’ve invited you to the **forked repo** and granted full access. You can push your refactor to the **[support-deepseek-coder-LLM](https://github.com/umans-tech/SWE-crafter/tree/support-deepseek-coder-LLM) branch** on _[umans-tech/SWE-crafter](https://github.com/umans-tech/SWE-crafter)_, and we'll see the changes in this PR.

Thanks!

jcraftsman commented on sweagent/agent/models.py:
```diff
@@ -257,6 +263,9 @@ def __init__(self, args: ModelArguments, commands: list[Command]):
                 api_version=keys_config.get("AZURE_OPENAI_API_VERSION", "2024-02-01"),
             )
+        elif self.args.model_name.startswith("deepseek") or self.args.model_name.startswith("coder"):
+            api_base_url: str | None = keys_config.get("DEEPSEEK_API_BASE_URL", None)
```
[Here](https://github.com/princeton-nlp/SWE-agent/pull/638/commits/0aa2956e87ad224d8971aadcecc9492c032b03b8) you go 🙂

klieret commented:
@jcraftsman Thanks! Could you check if it works with the last commit I've just pushed. After you give your OK, I'm happy to merge it :)

jcraftsman commented:
@klieret I really like your refactor, and everything continues to work perfectly. Thanks a lot! 🙂

klieret commented:
Thanks again for the PR! :)
